# Whole_exome_sequence

<h2>Data pre-processing</h2>

The pipeline follows the GATK4 (Genome Analysis Toolkit) [Best Practises](https://software.broadinstitute.org/gatk/best-practices), and [Germline short variant discovery (SNPs + Indels)](https://software.broadinstitute.org/gatk/best-practices/workflow?id=11145). The pipeline is designed under CentOS Linux 7.6.

Part of the Python script for visualize Insert size frequency and shell scipts for data selection are also included.
